### ISSUE TITLE
aio: service worker serves stale content on aio-staging

### ISSUE THREAD

**Issue by IgorMinar** [angular/angular#15198]
Repro steps:

1. If I go to https://aio-staging.firebaseapp.com I get the old version of the app:

  ![screen shot 2017-03-15 at 10 33 51 pm](https://cloud.githubusercontent.com/assets/216296/23983188/d7f7e5ba-09cf-11e7-8b15-412f78211af7.png)

2. As you can see in the screenshot, the main index.html was served by the service worker.

3. If I then reload the page, I still get the old stuff.

4. If I hard reload (cmd+shift+r) and bypass the service worker, I get the latest version:

  ![screen shot 2017-03-15 at 10 34 21 pm](https://cloud.githubusercontent.com/assets/216296/23983213/113027ac-09d0-11e7-829b-cee4105b8a6a.png)

5. If I "soft" reload again (cmd+r), I get the old version once again - served by the service worker.

What I expect is that the service worker will invalidate the cache once we push a new version to staging, but that's not currently happening and what's worse as a user I don't know how to invalidate the stale cache without going to devtools and removing it there.

-----

I'm also attaching the ngsw-manifest.json (served by the service-worker) in case that's helpful to debug this: 
[ngsw-manifest.json.txt](https://github.com/angular/angular/files/846776/ngsw-manifest.json.txt)

Is it expected that the manifest is served by the service worker? How is it supposed to get invalidated?

![screen shot 2017-03-15 at 10 42 54 pm](https://cloud.githubusercontent.com/assets/216296/23983356/c664e31a-09d0-11e7-95b3-81c956f2bb3c.png)

**Comment by alxhub** [comment]
I will look into this, but I think there is a good chance the behavior you're seeing is normal.

The NGSW update process works like this:

1) An update to the site is pushed.
2) The service worker script starts. This can happen when the user returns to an active tab after a period of inactivity, or they open the first application tab in their browser.
3) The SW downloads the manifest and learns of the available update.
4) It prepares for the update by downloading content that has changed hashes. 
5) The SW enters an UPDATE_PENDING state and continues serving requests with the old manifest.
6) At some point, the user navigates to the application with no other open tabs. This must be an actual first navigation, _not_ a refresh.
7) The SW detects that no other app tabs are open and transitions from UPDATE_PENDING to READY, setting the new manifest as active, and serves the new version of the app.

This process can get stuck at step 4, for example, if there are network or backend errors when downloading resources. In that case, the SW will retry the next time it starts up (caching only those requests which did not succeed the last time).

For insight into the SW's current state, http://aio-staging.firebaseapp.com/ngsw.log will show the current worker's lifecycle log, including the hashes of any manifests involved.

Also, yes, it's expected that the manifest is served by the service worker. This is an artifact of how the SW works - unless the manifest is listed in the manifest, the worker turns around and loads it from the network. In your screenshot you can see that the request for the manifest took ~400ms, as expected for a network request.

**Comment by alxhub** [comment]
I just walked AIO running locally through the update process.

First, the `ngsw.log` for the initial load:

```
{
  "state": "READY",
  "lifecycleLog": [
    "install event",
    "reset removed 0 ngsw: caches",
    "activate event",
    "transition from STARTUP to INSTALLING",
    "skipping update check, in state INSTALLING",
    "installed version a3fe4354a4c1913efac8079e903c760e20217e7f from network",
    "transition from INSTALLING to READY"
  ]
}
```

Then on a subsequent normal load:

```
{
  "state": "READY",
  "lifecycleLog": [
    "manifest a3fe4354a4c1913efac8079e903c760e20217e7f activated",
    "transition from STARTUP to READY"
  ]
}
```

Then I made a change to `main.ts` in AIO and stopped the service worker in dev tools (simulating inactivity in the app). On the next load, the SW saw the fresh manifest and entered `UPDATE_PENDING` as expected:

```
{
  "state": "UPDATE_PENDING",
  "lifecycleLog": [
    "manifest a3fe4354a4c1913efac8079e903c760e20217e7f activated",
    "transition from STARTUP to READY",
    "staged update to 7a9987124177882ded88b8164182e7e42fba78e8",
    "transition from READY to UPDATE_PENDING"
  ]
}
```

Finally, I stopped the SW and closed out all my open tabs. When I reopened the application the SW decided to serve the new version:

```
{
  "state": "READY",
  "lifecycleLog": [
    "manifest a3fe4354a4c1913efac8079e903c760e20217e7f activated",
    "staged manifest 7a9987124177882ded88b8164182e7e42fba78e8 present at initialization",
    "transition from STARTUP to UPDATE_PENDING",
    "skipping update check, in state UPDATE_PENDING",
    "updated to manifest 7a9987124177882ded88b8164182e7e42fba78e8",
    "transition from UPDATE_PENDING to READY",
    "cleaned up old version a3fe4354a4c1913efac8079e903c760e20217e7f"
  ]
}
```

I can't get the problem to reproduce locally, but it could be an artifact of Firebase hosting, so I'll continue to monitor AIO staging as changes go in and look for it to be in an inconsistent state.

I'll also work on some more granular debug logging in the SW so if we do have issues we have better visibility into why the update process failed.

**Comment by gkalpak** [comment]
(For reference: Related to #15561.)

**Comment by petebacondarwin** [comment]
@alxhub - any update on this?

**Comment by alxhub** [comment]
@petebacondarwin I believe @gkalpak's PR resolves this issue, do you agree?

**Comment by gkalpak** [comment]
Reading through the discussion again, I too believe that this specific issue should be resolved by #15561 (and friends). Closing...
